src/Admin/OrderAdminExtension.php

```diff
@@ -117,31 +116,31 @@ public function handleBatchPrintRequest($redirectTo, $action, $postIds)
         }
 
         if (!count($orderNumbers)) {
-            return esc_url(add_query_arg([
-                'pb_msg_type' => 'warning',
-                'pb_msg' => 'Please select some orders'
-            ], $redirectTo));
+            $this->message = 'Please select some orders';
+            $this->messageType = 'warning';
+            return $this->addNotificationQueryVars($redirectTo);
         }
 
         $printRequester = PrintRequesterFactory::create();
         try {
             $message = "Penny Black: " . $printRequester->printBatch($orderNumbers);
         } catch (PennyBlackException $e) {
-            return esc_url(add_query_arg([
-                'pb_msg_type' => 'error',
-                'pb_msg' => $e->getMessage()
-            ], $redirectTo));
+            $this->message = $e->getMessage();
+            $this->messageType = 'error';
+            return $this->addNotificationQueryVars($redirectTo);
         }
 
-        return esc_url(add_query_arg([
-            'pb_msg_type' => 'success',
-            'pb_msg' => $message
-        ], $redirectTo));
+        $this->message = $message;
+        $this->messageType = 'success';
+        return $this->addNotificationQueryVars($redirectTo);
     }
 
     public function addNotificationQueryVars($location)
     {
-        return esc_url(add_query_arg(['pb_msg_type' => $this->messageType, 'pb_msg' => $this->message], $location));
+        return add_query_arg(
+            ['pb_msg_type' => rawurlencode($this->messageType), 'pb_msg' => rawurlencode($this->message)],
```

Comment:
`esc_url()` was wrong, this creates a whole URL for you, even if what you pass isn't a URL, so it's not usable for URL _parameters_. Switched to just `rawurlencode` and updated all approaches to use this method.